### ISSUE TITLE
adams2019 autoscheduler compilation failure on ppc64le

### ISSUE THREAD

**Issue by cimes-isi** [halide/Halide#5571]
I'm trying to build Halide on ppc64le and am now on commit 10a01dd.  I'm using a module for LLVM 11.0.0-rc1 on the Summit supercomputer.

```
[ 52%] Generating train_cost_model.h, train_cost_model.o
cd /ccs/home/cimes/casper/Halide-master/build-llvm11rc1/src/autoschedulers/adams2019 && ./cost_model.generator -n train_cost_model -d 0 -g train_cost_model -f train_cost_model -e c_header,object -o . target=powerpc-64-linux-no_runtime
Call parameter type does not match function signature!
  %55 = shufflevector <8 x float> %54, <8 x float> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
 <8 x float>  %56 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vmaxfp(<4 x float> %55, <4 x float> zeroinitializer) #5
Call parameter type does not match function signature!
  %57 = shufflevector <8 x float> %54, <8 x float> undef, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
 <8 x float>  %58 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vmaxfp(<4 x float> %57, <4 x float> zeroinitializer) #5
Call parameter type does not match function signature!
  %142 = shufflevector <8 x float> %141, <8 x float> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
 <8 x float>  %143 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vmaxfp(<4 x float> %142, <4 x float> zeroinitializer) #5
Call parameter type does not match function signature!
  %144 = shufflevector <8 x float> %141, <8 x float> undef, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
 <8 x float>  %145 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vmaxfp(<4 x float> %144, <4 x float> zeroinitializer) #5
Call parameter type does not match function signature!
  %210 = shufflevector <8 x float> %209, <8 x float> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
 <8 x float>  %211 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vmaxfp(<4 x float> %210, <4 x float> <float 1.000000e+00, float 1.000000e+00, float 1.000000e+00, float 1.000000e+00>) #5
Call parameter type does not match function signature!
  %212 = shufflevector <8 x float> %209, <8 x float> undef, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
 <8 x float>  %213 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vmaxfp(<4 x float> %212, <4 x float> <float 1.000000e+00, float 1.000000e+00, float 1.000000e+00, float 1.000000e+00>) #5
Call parameter type does not match function signature!
  %285 = shufflevector <8 x float> %t15898.s, <8 x float> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
 <8 x float>  %286 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vmaxfp(<4 x float> %285, <4 x float> <float 1.000000e+00, float 1.000000e+00, float 1.000000e+00, float 1.000000e+00>) #5
Call parameter type does not match function signature!
  %287 = shufflevector <8 x float> %t15898.s, <8 x float> undef, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
 <8 x float>  %288 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vmaxfp(<4 x float> %287, <4 x float> <float 1.000000e+00, float 1.000000e+00, float 1.000000e+00, float 1.000000e+00>) #5

<snipped>

Call parameter type does not match function signature!
  %159 = shufflevector <8 x i32> %155, <8 x i32> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
 <8 x i32>  %161 = call <4 x i32> @llvm.ppc.altivec.vminsw(<4 x i32> %159, <4 x i32> %160) #5
Call parameter type does not match function signature!
  %162 = shufflevector <8 x i32> %155, <8 x i32> undef, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
 <8 x i32>  %164 = call <4 x i32> @llvm.ppc.altivec.vminsw(<4 x i32> %162, <4 x i32> %163) #5
Call parameter type does not match function signature!
  %45 = shufflevector <8 x float> %44, <8 x float> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
 <8 x float>  %46 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vminfp(<4 x float> %45, <4 x float> zeroinitializer) #5
Call parameter type does not match function signature!
  %47 = shufflevector <8 x float> %44, <8 x float> undef, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
 <8 x float>  %48 = call reassoc nnan ninf nsz contract afn <4 x float> @llvm.ppc.altivec.vminfp(<4 x float> %47, <4 x float> zeroinitializer) #5
Unhandled exception: Internal Error at /ccs/home/cimes/casper/Halide-master/src/CodeGen_LLVM.cpp:687 triggered by user code at : Condition failed: !verifyModule(*module, &llvm::errs()): 

make[2]: *** [src/autoschedulers/adams2019/train_cost_model.h] Error 255
make[2]: Leaving directory `/autofs/nccs-svm1_home1/cimes/casper/Halide-master/build-llvm11rc1'
make[1]: *** [src/autoschedulers/adams2019/CMakeFiles/train_cost_model.update.dir/all] Error 2
make[1]: Leaving directory `/autofs/nccs-svm1_home1/cimes/casper/Halide-master/build-llvm11rc1'
make: *** [all] Error 2
```

**Comment by Infinoid** [comment]
I don't have access to a POWER9 supercomputer, but I am able to reproduce this error using qemu.  I can provide setup instructions if needed.

**Comment by cimes-isi** [comment]
I will also note that we did successfully build Halide v10.0.0 on this platform.

**Comment by steven-johnson** [comment]
> I don't have access to a POWER9 supercomputer, but I am able to reproduce this error using qemu.

Definite Quote Of The Day material

**Comment by cimes-isi** [comment]
LOL, didn't mean to sound elitist.  In fact, @Infinoid produced this problem in QEMU a week or so ago, before I saw it on real hardware.  We thought it might have been a QEMU quirk so didn't pursue it at the time.  Hopefully this is an indication that QEMU might be a useful test environment for less common architectures.

**Comment by Infinoid** [comment]
If you want to do the QEMU thing, [here](https://gist.github.com/Infinoid/62d28559aebeddb7d3d53fe4418e5dc4) are some setup instructions, and the scripts I used.  There's a link to some logs at the bottom of the README.md file.

It's user-level emulation, meaning each Linux process gets its own copy of qemu, and it's just translating ppc64le instructions and passing system calls to your host kernel.

**Comment by bird-rookie** [comment]
hi，cimes-isi 。
I'm a beginner of adams2019 autoscheduler。
I wonder how can i run the autoscheduler in linux with the master code, there is no README and come to some problem.
can you tel me, thank you so much.

**Comment by Infinoid** [comment]
> hi，cimes-isi 。
> I'm a beginner of adams2019 autoscheduler。
> I wonder how can i run the autoscheduler in linux with the master code, there is no README and come to some problem.
> can you tel me, thank you so much.

If you are building things by hand or with a hand-written Makefile, I would start by looking at [Tutorial 21](https://halide-lang.org/tutorials/tutorial_lesson_21_auto_scheduler_generate.html).  Make sure that works for you, then replace `mullapudi2016` with `li2018` or `adams2019` to select other autoschedulers.

If you are using cmake, there is a note about it in [the cmake readme](https://github.com/halide/Halide/blob/master/README_cmake.md#user-content-autoschedulers), and you can also look at how [the example apps](https://github.com/halide/Halide/blob/master/apps/bilateral_grid/CMakeLists.txt#L27) do it.

If you are having specific problems, try posting a new issue about it or ask on [the gitter channel](https://gitter.im/halide/Halide) - there are many helpful people in there.

**Comment by cimes-isi** [comment]
Happy new year!

I'm now using a custom LLVM 11.0.0 build to get past some other issues on the platform, and I've moved my master branch forward to revision 8063879.  Unfortunately it's still failing at this point.  The output is a little different, but perhaps more readable since it's not all lowered LLVM instructions:

```
[ 50%] Generating cost_model.h, cost_model.o
cd /ccs/home/cimes/casper/Halide-master/build-llvm-20210106/src/autoschedulers/adams2019 && ./cost_model.generator -n cost_model -d 0 -g cost_model -f cost_model -e c_header,object -o . target=powerpc-64-linux-no_runtime
cost_model.generator: /ccs/home/cimes/casper/llvm-project-11/llvm/lib/IR/Instructions.cpp:453: void llvm::CallInst::init(llvm::FunctionType*, llvm::Value*, llvm::ArrayRef<llvm::Value*>, llvm::ArrayRef<llvm::OperandBundleDefT<llvm::Value*> >, const llvm::Twine&): Assertion `(i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!"' failed.
/bin/sh: line 1:  2828 Aborted                 (core dumped) ./cost_model.generator -n cost_model -d 0 -g cost_model -f cost_model -e c_header,object -o . target=powerpc-64-linux-no_runtime
make[2]: *** [src/autoschedulers/adams2019/cost_model.h] Error 134
make[2]: Leaving directory `/autofs/nccs-svm1_home1/cimes/casper/Halide-master/build-llvm-20210106'
make[1]: *** [src/autoschedulers/adams2019/CMakeFiles/cost_model.update.dir/all] Error 2
make[1]: Leaving directory `/autofs/nccs-svm1_home1/cimes/casper/Halide-master/build-llvm-20210106'
make: *** [all] Error 2
```

**Comment by cimes-isi** [comment]
Probable duplicate of #5632, fixed by #5531.
